### ISSUE TITLE
When staging a selection of files, wrong files are sometimes staged

### ISSUE THREAD

**Issue by nyanpasu64** [git-cola/git-cola#1117]
When I select multiple files and stage or unstage, sometimes not all files get staged or unstaged. The effects are more obvious if there are multiple categories of files (Untracked, Modified, or Staged).

https://user-images.githubusercontent.com/913957/110625447-47fe0d80-8154-11eb-8f7d-5703af48c3fb.mp4

One easy and deterministic way to reproduce this issue is:

- Click a heading (Untracked, Modified, or Staged)
- Optionally Ctrl+A.
- Stage or unstage (either Ctrl+S to toggle staging, or click the Stage and/or Unstage button).
- Immediately Ctrl+A to select all files.

One second later, the Diff panel will update, and (if you pressed Ctrl+A after staging/unstaging), the Status list of files will deselect all files except for the headings. At this point, if you have multiple headings and multiple categories of files, pressing Ctrl+S again will only affect one category (files under one heading).

https://user-images.githubusercontent.com/913957/110625420-403e6900-8154-11eb-9e85-0b8b326186b0.mp4

The results of this bug are very unpredictable.

## Version

- OS: Arch Linux
- Version: Arch Linux AUR, both 3.9-2 and Git 3359dc7c2651ade666c41dd3daf7563ab9815685.
- Qt: System Qt (not PyQt5 wheel), qt5-base 5.15.2-5 (built myself).

I think I've gotten bugs like this for years, but only recently took the time to investigate the exact behavior.

**Comment by davvid** [comment]
I think we can probably add more special cases to make this better behaved.

We have logic in the selection -> action code that attempts to save the user from weird situations.

One weird situation is when you have a file that is partially staged, so it shows up in both the Modified and Staged sections. If we select that file in both categories and then do ctrl-s then it's ambiguous what to do.

The internal logic is that it looks at the various different categories (staged, modified, unmerged, untracked) and only uses the first one in a priority-list order that has selections in that category to try to do something simple. Anyways, that's implementation details, and what you're showing here is that this abstraction has some places where we can make improvements.

In the described situation it'll only perform the unstaging of the staged files, and nothing more. I think we can improve this behavior for the situation in the video, though.

In the video we have a selection that has modified and untracked files only. That's a scenario we can special-case and make it stage 'em all.

Until then, please try using the following hotkeys if you're looking to quickly stage everything.

`alt-a` will stage all of the modified files.

`alt-u` will stage all of the untracked files.

That's the same number of keystrokes as `ctrl-a ctrl-s` so hopefully it helps out.

Another note is that we don't have a hotkey at `ctrl-alt-a` which could perhaps stage modified and untracked files in one shot. Should we add it?

**Comment by davvid** [comment]
Alrighty, this led to a few improvements all around. The status tree should now behave much more predictably. Doing `ctrl-a ctrl-s` will now behave as expected.  We now process the entire selection and only special-case the scenario where the Staged header is part of the selection, so it should be much better behaved now.

Also, I noticed that we didn't have a really easy one-shot way to stage or unstage everything (both modified + untracked), so the new `ctrl + shift + s` hotkey is now available for this purpose.  If files are staged, they'll get unstaged, else if files are unstaged (modified + untracked) then they will all get staged.

Please test the main branch if you can.  Thanks again for the report!

**Comment by nyanpasu64** [comment]
The wonky staging/unstaging behavior seems to be gone. The status selection and diff display still changes 1 second after a staging operation though.

Didn't know that Alt+A, Alt+U (and the newly added Ctrl+Shift+S) even existed. Apparently the actions and shortcuts are exposed in the Commit menu (which I didn't check often), but not when you right-click a file in the Status panel (which I did). Do you think it's a good idea to expose those operations in the right-click menu?

Also all icons are missing (the status arrow icons, and the settings checkboxes). I don't know if it's an Arch/Qt bug or a git-cola bug (happens in both AUR and `pip install .`). I'll investigate this further, since it makes it hard to tell what files I'm staging and unstaging.

**Comment by davvid** [comment]
The best way to see the shortcuts is to hit `?`.  It'll popup the shortcuts cheat sheet. ctrl-j/k/l and j/k are my favorites.

We don't (currently) support `pip install` as the installation method. That might be why the icons are missing.

If you're looking to install to a particular location (eg. `$HOME/.local` or a virtualenv) you can do `make install prefix=<path>` to install to the specified path.

This is virtualenv compatible. Specify the virtualenv as the `make install prefix=...` directory while the virtualenv is activate and it'll get installed into the virtualenv. The installation will then be self-contained within the virtualenv directory.

You don't need to install anything if you have a git clone, though. Run `python git-cola/bin/git-cola` to run in-place.

Thanks for the note about the delayed update. This could have to do with the size of the diff, but it's a guess without seeing it live. Is there any chance this is reproducible using a public repo?

Nonetheless, one way to see which (git) operations might be taking the longest is to run `GIT_COLA_TRACE=1 ...` where `...` how you'd normally run cola. It'll print out all of the git commands it runs after they finish, so if it takes long before a certain command is printed then that command is the slow one.

We don't currently include timing information in the trace output. That would be good to add.

**Comment by nyanpasu64** [comment]
Both `python bin/git-cola` and the AUR git-cola-git package (built through make/install, not pip) are missing UI icons. However the icons are properly installed in `/usr/share/git-cola/icons/a-z-order.svg` etc. So I don't know what's causing the issue. Hopefully I'll be able to reinstall Linux (a different distro) when I get a new computer, and hopefully my OS will be less of a buggy mess then.

I've encountered the delayed update issue when working with the https://gitlab.com/exotracker/exotracker-cpp repo. I checked out the (default) dev branch, then `git checkout --detach` and `git reset master` to generate a large number of unstaged changes. Afterwards, I ran `GIT_COLA_TRACE=1 git-cola`:

<blockquote><details><summary>Logs from git-cola</summary>

```
(Stage or unstage a file)

git -c diff.suppressBlankEmpty=false -c log.showSignature=false check-attr encoding -- src/gui/pattern_editor.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff -M --cached --patience --submodule --no-color --no-ext-diff --unified=3 HEAD -- src/gui/pattern_editor.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false check-attr encoding -- 3rdparty/Blip_Buffer/Blip_Buffer.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff -M --patience --submodule --no-color --no-ext-diff --unified=3 -- 3rdparty/Blip_Buffer/Blip_Buffer.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false add --force --verbose -- 3rdparty/Blip_Buffer/Blip_Buffer.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff-index --cached -z HEAD --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff-files -z --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false ls-files -z --others --exclude-standard --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false rev-parse --symbolic-full-name HEAD
git -c diff.suppressBlankEmpty=false -c log.showSignature=false check-attr encoding -- 3rdparty/verdigris/wobjectimpl.h
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff -M --patience --submodule --no-color --no-ext-diff --unified=3 -- 3rdparty/verdigris/wobjectimpl.h

(1 second passes)

git -c diff.suppressBlankEmpty=false -c log.showSignature=false update-index --refresh
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff-index --cached -z HEAD --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff-files -z --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false ls-files -z --others --exclude-standard --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false rev-parse --symbolic-full-name HEAD
git -c diff.suppressBlankEmpty=false -c log.showSignature=false remote
git -c diff.suppressBlankEmpty=false -c log.showSignature=false for-each-ref --format=%(refname) --sort=version:refname
git -c diff.suppressBlankEmpty=false -c log.showSignature=false rev-parse --symbolic-full-name HEAD
git -c diff.suppressBlankEmpty=false -c log.showSignature=false submodule status
git -c diff.suppressBlankEmpty=false -c log.showSignature=false ls-files -z --
git -c diff.suppressBlankEmpty=false -c log.showSignature=false diff -M --patience --submodule --no-color --no-ext-diff --unified=3 -- 3rdparty/verdigris/wobjectimpl.h
```

</details></blockquote>

**Comment by richarson** [comment]
@nyanpasu64 BTW, about icons not showing up, are you using KDE Plasma? If so, there was a bug in KFrameworks 5.80 regarding svg icons, it's fixed in KF >= 5.81.

**Comment by nyanpasu64** [comment]
Yeah I ended up noticing the icons being fixed, didn't realize it was due to a KF5.80 bug.
